### PR TITLE
Add an option to colorize to respect the destination alpha

### PR DESCRIPTION
**Pull request by ssieb**
Also rework the colorizing code to be more efficient.

### CODE REVIEW

**Comment by ssieb** [comment]
I also updated the doc file to better explain how the colorize function works.

**Comment by ssieb** [comment]
The current colorize code ignores the alpha value of the texture being shaded.  As long as it is not 0, it gets replaced with the color specified, including the alpha value.  This makes it difficult to create shaded textures with arbitrary colors.  I was able to work around this by making sure the texture I was coloring used only white, then used it again as a mask to fix the alpha values.

The new code I've added here adds an option to the colorize command to multiply the alpha of the color by the alpha of the texture.  This preserves the opacity of the texture and keeps the shading.

I looked into various mods I have installed and it looks like changing this for all uses would affect their appearance.  That's why I added a new parameter.  I don't really care what the parameter is, as long as it is available.  Using "alpha" in the ratio slot means that it will only have an effect on clients with this change.  Any older clients will just ignore it and will still look like the before picture.

Before: can_before
After: can_after

These pictures show another issue with the way the current code handles alpha.  It took me quite a while to figure out where those extra bits at the top of the brushes came from.  Those pixels are almost entirely transparent, but not quite, so they are replaced with full color.

**Comment by ObaniGemini** [comment]
Wow great for not using many different textures :D

**Comment by kwolekr** [comment]
Great work.  The end result looks nice, and everybody appreciates the optimization.

One thing though, the code style dictates that there must be spaces between operators, like `foobar / 255.0f`, not `foobar/255.0f`.

File: src/client/tile.cpp [hunk 1]
```diff
@@ -1784,6 +1781,46 @@ static void blit_with_interpolate_overlay(video::IImage *src, video::IImage *dst
...
+				}
+			}
+		} else {
+			for (u32 y = dst_pos.Y; y < dst_pos.Y + size.Y; y++)
+			for (u32 x = dst_pos.X; x < dst_pos.X + size.X; x++)
+				if (dst->getPixel(x, y).getAlpha() > 0)
+					dst->setPixel(x, y, color);
+		}
+	} else {
+		float interp = (ratio == -1 ? (float)color.getAlpha()/255.0f : (float)ratio/255.0f);
```
Comment (kwolekr) [inline src/client/tile.cpp]: These casts aren't necessary.

File: src/client/tile.cpp [hunk 2]
```diff
@@ -1784,6 +1781,46 @@ static void blit_with_interpolate_overlay(video::IImage *src, video::IImage *dst
...
+		video::SColor color, int ratio, bool keep_alpha)
+{
+	u32 alpha = color.getAlpha();
+	video::SColor dst_c;
+	if (alpha == 0)
+		return;
+	if (((ratio == -1) && (alpha == 255)) || (ratio == 255)) {
+		if (keep_alpha) {
+			dst_c = color;
+			for (u32 y = dst_pos.Y; y < dst_pos.Y + size.Y; y++)
```
Comment (kwolekr) [inline src/client/tile.cpp]: These coordinates are unsigned whereas the v2s32 struct members are signed.

File: src/client/tile.cpp [hunk 3]
```diff
@@ -1784,6 +1781,46 @@ static void blit_with_interpolate_overlay(video::IImage *src, video::IImage *dst
...
 
 /*
+	Apply color to destination
+*/
+static void apply_colorize(video::IImage *dst, v2s32 dst_pos, v2u32 size,
+		video::SColor color, int ratio, bool keep_alpha)
+{
+	u32 alpha = color.getAlpha();
+	video::SColor dst_c;
+	if (alpha == 0)
```
Comment (kwolekr) [inline src/client/tile.cpp]: Why the special case for a 0 alpha?

File: src/client/tile.cpp [hunk 4]
```diff
@@ -1784,6 +1781,46 @@ static void blit_with_interpolate_overlay(video::IImage *src, video::IImage *dst
...
+		return;
+	if (((ratio == -1) && (alpha == 255)) || (ratio == 255)) {
```
Comment (kwolekr) [inline src/client/tile.cpp]: This might be a little less visually confusing if there were less ( (( )) )) parentheses.

**Comment by kwolekr** [comment]
:+1:

**Comment by HybridDog** [comment]
l'm not sure if it here's the right place to write this:
colorize texture modifiers break normalmaps

**Comment by kwolekr** [comment]
This is the wrong place for it.  This is a PR for adding an option to the existing colorize and can't break anything.  Perhaps you should create a new issue in github for an existing problem you've experienced.

**Comment by kwolekr** [comment]
https://github.com/minetest/minetest/commit/01ae43c48009f816f4649fae2f7f6997452aa6cf

**Comment by est31** [comment]
Who gave the second approval?

**Comment by celeron55** [comment]
@kwolekr I don't think that was appropriate; you merged a PR that had tens of lines of diff AND includes an API modification, solely by yourself, based on an IRC discussion where you discussed with no other core developers and only with the contributor of said PR.

**Comment by paramat** [comment]
Sorry celeron55, there has been some confusion about feature freeze, it's not in full effect yet but we are mostly focussing on PRs essential or desirable for the release, see https://forum.minetest.net/viewtopic.php?f=18&t=14102&start=50#p211876

**Pull request closed without merging**